### PR TITLE
Use LCS for wire transfer of SignedTransaction

### PR DESCRIPTION
**Pull request by davidiw**
There are a lot of commits here:
- Make RawTransaction use LCS for hashing / signing
- Eliminate RawTransactionBytes and merge the functions into RawTransaction (no longer needed with LCS implementation)
- Cleanup wallet usage of keys to make it work better in our crypto::ed25519 world
- Eliminate proto code for RawTransaction and substantially simplify it for SignedTransaction

I'm making this a single PR and am happy to move around the commits.

### CODE REVIEW

File: types/src/proto/transaction.proto [hunk 1]
```diff
@@ -73,16 +43,8 @@ message Module {
...
-    // across different language implementations, but for our use cases for
-    // transaction it is not necessary because the client is the only one to
-    // produce this bytes, which is then persisted in storage.
-    bytes raw_txn_bytes = 1;
-    // public key that corresponds to RawTransaction::sender_account
-    bytes sender_public_key = 2;
-    // signature for the hash
-    bytes sender_signature = 3;
+    // LCS byte code representation of a SignedTransaction
+    bytes signed_txn = 5;
```
Comment (weilei0107) [inline types/src/proto/transaction.proto]: 5?

File: types/src/transaction.rs [hunk 2]
```diff
@@ -471,17 +417,20 @@ impl fmt::Debug for SignedTransaction {
 }
 
 impl SignedTransaction {
-    pub fn craft_signed_transaction_for_client(
+    pub fn new(
```
Comment (weilei0107) [inline types/src/transaction.rs]: a nit that if we rename it to new the return type is usually 'Self'

File: types/src/transaction.rs [hunk 3]
```diff
@@ -560,43 +509,20 @@ impl CryptoHash for SignedTransaction {
 impl FromProto for SignedTransaction {
```
Comment (weilei0107) [inline types/src/transaction.rs]: Do we still need this?

File: types/src/proto/transaction.proto (hunk 1 above)
Comment (davidiw) [reply types/src/proto/transaction.proto]: we should reset this to 0 at some point in time.... 4 was used in an early commit in this stack.

**Comment by davidiw** [comment]
> There's a reason that we do not use the new crypto API in the wallet, notably that we want a clean separation of dependencies, for this project that we think should be distributed as its own binary.
> 
> This pulls in too much of the code base. While I'd be happy with it adopting the principles of the crypto API, I'd like to see that not require pulling in BLS or X5519 first.

The wallet code leverages the SignedTransaction type. The SignedTransaction type depends on the crypto::ed25519, hence these dependencies will either be pulled in either explicitly or implicitly.

**Comment by kphfb** [comment]
I'm a bit mixed on this.  While I understand @huitseeker's concern, it also is already being pulled in via SignedTransaction.  So if we actually wanted to limit this, we would need to remove that dependency first which is a bigger change that would be unrelated to this diff.  I think we should allow this one to move forward to unblock the change and can think about removing the dependency later if we desire, although I have a feeling that is going to be a larger change that we may not wish to make

**Comment by sunshowers** [comment]
@davidiw -- noticed that this removes the `raw_transaction` fuzzing target. Is that because fuzzing `signed_transaction` would cover `raw_transaction` as well?

**Comment by davidiw** [comment]
> @davidiw -- noticed that this removes the `raw_transaction` fuzzing target. Is that because fuzzing `signed_transaction` would cover `raw_transaction` as well?

Yes, we no longer write raw transactions directly. We can go further to remove SignedTransaction proto as well, which would require we delete that as well. We could still build fuzzing for SubmitTransactionRequest to cover all cases though.

**Comment by sunshowers** [comment]
Hmm, maybe it's worth fuzzing the LCS deserializers instead of removing the fuzz targets? :)

**Comment by mimoo** [comment]
I think it's a good idea to fuzz the LCS deserializer for raw transactions. If you push this commit I can work on this.

**Pull request merged**